src/natcap/invest/validation.py

```diff
@@ -930,65 +929,48 @@ def validate(args, spec, spatial_overlap_opts=None):
-    # it has a value.  A sufficient input need not be valid.  Sufficiency is
-    # used by the conditional requirement phase (step 3 in this function) to
-    # determine whether a conditionally required input is required.
-    # The only special case about sufficiency is with boolean values.
-    # A boolean value absent from args is insufficient.  A boolean input that
-    # is present in args but False is in sufficient.  A boolean input that is
-    # present in args and True is sufficient.
+    # has a truthy value.  If the input is missing from args or is falsy, it is
+    # insufficient.
     insufficient_keys = missing_keys.union(keys_with_no_value)
     sufficient_inputs = {}
-    for key, parameter_spec in spec.items():
-        # If the key isn't present, no need to validate.
-        # If it's required and isn't present, we wouldn't have gotten to this
-        # point in the function.
-        if key not in args:
+    insufficient_keys = set()
+    for key in spec.keys():
+        try:
+            sufficient_inputs[key] = bool(args[key])
+        except KeyError:
+            # input is definitely insufficient if it's missing from args
             sufficient_inputs[key] = False
-            insufficient_keys.add(key)
-            continue
 
-        # If the value is empty and it isn't required, then we don't need to
-        # validate it.
-        if args[key] in ('', None):
-            sufficient_inputs[key] = False
+        if sufficient_inputs[key] is False:
             insufficient_keys.add(key)
-            continue
-
-        # Boolean values are special in that their T/F state is equivalent
-        # to their satisfaction.  If a checkbox is checked, it is
-        # considered satisfied.
-        if spec[key]['type'] == 'boolean':
-            sufficient_inputs[key] = args[key]
-
-        # Any other input type must be sufficient because it is in args and
-        # has a value.
-        else:
-            sufficient_inputs[key] = True
 
-    # step 3: evaluate required status of conditionally required keys
-    # keep track of keys that are explicity not required due to
-    # their condition being false
+    # Step 3: evaluate whether conditionally required keys are required.
+    # We also want to keep track of keys that are not required due to their
+    # conditional requirement expression evaluating to False.
+    #
+    # An input is conditionally required when the expression given is
+    # truthy.
+    expression_values = {
+        input_key: args.get(input_key, False) for input_key in spec.keys()}
     excluded_keys = set()
     for key in conditionally_required_keys:
-        # An input is conditionally required when the expression given
-        # evaluates to True.
         is_conditionally_required = _evaluate_expression(
-            expression=spec[key]['required'],
-            variable_map=sufficient_inputs)
+            expression=f'bool({spec[key]["required"]})',
```

Comment:
This is a really minor stylistic thing, but I wonder would it make sense to cast the result of `_evaluate_expression` to `bool`, rather than putting `bool(...)` in the expression? Just to minimize the amount of code that we have to `eval`

Reply:
Makes sense to me!  It'll be a little less to re-`eval` on each call, so there's probably a negligible speedup this way too.  Patched (and added a separate missing cast to `bool` in d70b15dae)